web/src/engine/osk/src/visualKeyboard.ts

```diff
@@ -488,42 +502,41 @@ export default class VisualKeyboard extends EventEmitter<EventMap> implements Ke
         // First, if we've configured the gesture to generate a keystroke, let's handle that.
         const gestureKey = gestureStage.item;
 
-        let coordSource = gestureStage.sources[0];
-        let coord: InputSample<KeyElement, string> = null;
-        if(coordSource) {
-          // TODO:  should probably vary depending upon `gestureStage.matchedId`
-          // (certain types should probably use the base coord... or even from
-          // a prior stage of the sequence as appropriate.)
-          //
-          // This is the coordinate used as the basis for fat-finger calculations.
-          coord = coordSource.currentSample;
-        }
+        const coordSource = gestureStage.sources[0];
+        const coord: InputSample<KeyElement, string> = coordSource ? coordSource.currentSample : null;
 
-        let keyResult: {
-          contextToken?: number
-        } = null;
+        let keyResult: KeyRuleEffects = null;
 
         // Multitaps do special key-mapping stuff internally and produce + raise their
         // key events directly.
-        if(gestureKey && !(handler instanceof Multitap)) {
+        if(gestureKey && !(handlers && handlers[0] instanceof Multitap)) {
           let correctionKeyDistribution: KeyDistribution;
-          const baseDistanceMap = this.getSimpleTapCorrectionDistances(coord, gestureKey.key.spec as ActiveKey);
+          const baseDistanceMap = this.getSimpleTapCorrectionDistances(coordSource.currentSample, gestureKey.key.spec as ActiveKey);
 
-          if(handler) {
+          if(handlers) {
             // Certain gestures (especially flicks) like to consider the base layout as part
             // of their corrective-distribution calculations.
             //
             // May be `null` for gestures that don't need custom correction handling,
             // such as modipresses or initial/simple-tap keystrokes.
-            correctionKeyDistribution = handler.currentStageKeyDistribution(baseDistanceMap);
+            correctionKeyDistribution = handlers[0].currentStageKeyDistribution(baseDistanceMap);
           }
 
           if(!correctionKeyDistribution) {
             correctionKeyDistribution = distributionFromDistanceMaps(baseDistanceMap);
           }
 
           // Once the best coord to use for fat-finger calculations has been determined:
-          keyResult = this.modelKeyClick(gestureStage.item, coord);
+          const shouldLockLayer = handlers && (handlers[0] instanceof SubkeyPopup) && handlers[0].shouldLockLayer;
+          try {
+            if(shouldLockLayer) {
+              this.lockLayer(true);
+            }
+            keyResult = this.modelKeyClick(gestureStage.item, coord);
+          } finally {
+            this.lockLayer(false);
+          }
```

Comment:
Why do we need `try` here? Are we expecting exceptions to be thrown?

Reply:
Expecting?  No, not during standard use.   But that's why they're "exceptions".

`this.modelKeyClick` triggers code that links into rule processing, which isn't exactly a narrow area of code.  Should something unexpectedly break, we want to at least "unlock" the layer even if an error occurs... otherwise, the OSK will appear to "lock up" to end users, and that's far from ideal.

Sometimes we've seen errors from individual keystrokes that don't break things for other keystrokes.  If KMW isn't completely broken, just failing that one keystroke can allow a form of recovery... in which we want to ensure that any layer-lock (modipress) manipulations we do here also recover if possible.

Reply:
Okay, we just haven't got anywhere near that level of robustness elsewhere in KeymanWeb, so this is a bit of a new pattern.

Reply:
A text-search across the repo for `} finally {` actually turns up 8 pre-existing instances of this exact pattern.